cmd/agent_wrapper/main_linux.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !windows
+
+package main
+
+import (
+	"log"
+	"os"
+	"os/exec"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+func handleSignals(cmd *exec.Cmd) {
+	sigs := make(chan os.Signal, 1)
+	for {
+		signal.Notify(sigs, syscall.SIGTERM, syscall.SIGINT)
+		sig := <-sigs
+		start := time.Now()
+		for {
+			if cmd.Process != nil {
+				cmd.Process.Signal(sig)
+				break
+			} else if 80*time.Second+time.Until(start) <= 0 {
+				break
+			}
+			time.Sleep(50 * time.Millisecond)
+		}
+		if sig == syscall.SIGTERM {
+			time.Sleep(80*time.Second + time.Until(start))
+			log.Fatalf("Wrapped %v did not terminate in 80 seconds\n", cmd.Path)
+		}
+	}
```

Comment:
I think I'm confused about a few things:

* What's the purpose of the outer `for` loop? `sig := <-sigs` will block until a signal is received, and all of the signals we subscribe to are terminating.
* Why is `signal.Notify` inside the `for` loop? I can't think of a reason why we'd need to subscribe the same signals to the same channel more than once.
* The `did not terminate` error doesn't make sense to me -- it seems we're actually trying to check whether the process was _started_, not terminated.
  * In that case, why is this a fatal error? We can still handle this situation gracefully in the wrapper.
  * Why are we checking for `SIGTERM` on line 43? The condition for "process was/wasn't started" should be the result of the `cmd.Process != nil` evaluation on line 35 and should have nothing to do with whether the signal was `SIGTERM` or `SIGINT`.
* What's the reason for the `time.Sleep` before `log.Fatalf`?

Reply:
1. I was missing the SIGCONT signal that fluent-bit uses to dump information, updated it
2. Good catch, thanks!
3. After thinking about it, I think that the `did not terminate` block is not required.  It was meant to make sure the sub-process doesn't take too long to stop after receiving a SIGTERM.
  3.2 I added some comments to explain why we check for `cmd.Process != nil`. It is unrelated to what I was trying to do with the `process did not terminate` block
5. It was meant to give the subprocess time to stop before being killed, as it needs time to do cleanup

Reply:
Makes sense, thanks for the clarification!